### DISCUSSION TITLE
Pricing Questions: Can I configure next-right-now to be FREE?

### DISCUSSION THREAD

**Discussion by joshualinog** [UnlyEd/next-right-now#16] [Q&A]
# I'm experimenting....

After a cursory reading and watching your 1.5 of your videos, It seems locize is the only requirement that has no free plan.

## Currently I don't need ~~locize~~ a few libs/utils, but would love to stick as closely to your practices
I know you said it was _opinionated_, but I am kind of hoping to use your approaches/methodologies like it's my web-dev-rabbi/teacher. 

> NRN is strongly opinionated, but also very flexible. It can also be used as a learning/teaching resource. **_-dude who wrote the readme_**

![image](https://user-images.githubusercontent.com/1839603/77454136-8200c300-6dce-11ea-9c47-1f5cbabd5b2b.png)

## I'm broke
I would like to keep the project I build free for as long as possible, **WHILST STICKING CLOSELY TO YOUR PRACTICES**

# EDITS
It actually seems GraphCMS is also lacking a "truly" free tier, as they limit content entries (any entry whatsoever) to 500. 

## I see the removal section, but not GraphCMS as an option...
https://github.com/UnlyEd/next-right-now/blob/master/README_HOW_TO_REMOVE.md

**Comment by Vadorequest** [comment]
@jajourda  Hello! I understand your pain.

About Locize, they may have some discount but if you don't need them in the first place then it's no use to you. The best solution I see here is to remove Locize from your project.

I was considering yesterday to release some **variant** versions, without Locize for instance. or without GraphCMS, etc. I believe you could use those. 😉 

GraphCMS has some free plan but it's limited indeed. There is also a 2weeks trial on Growth plan.
The good news is, you don't need GraphCMS, you can use any GraphQL backend you'd like. If you're looking for an OSS alternative I recommend Strapi http://strapi.io/

Using strapi, you could run your graphql backend on your computer and play around locally for free. Hosting strapi online would cost you money though, as it needs at least 2Go RAM which requires a VPS (or similar, like AWS EC2, etc.), or you could host it yourself but that requires some knowledge about networks and such, which aren't covered here.

Or find another GraphQL server free alternative, I don't know any but there is likely some that exists.

I'd be interested to speak with you regarding the "variants" I mentioned above. If you'd like to, please contact me at https://twitter.com/Vadorequest :)

**Comment by joshualinog** [comment]
# thoughts on Sanity?
You recommended Strapi, but what about sanity?

https://www.sanity.io/graphql

## supposedly a nextjs example
https://www.sanity.io/create?template=sanity-io/sanity-template-nextjs-landing-pages

**Comment by Vadorequest** [comment]
Because their GraphQL is actually beta:
https://www.sanity.io/docs/what-about-graphql

I had studied their system and they were using 2 different "ways" for queries and mutations, which didn't use the same system. Makes things harder to do. It may be a viable alternative, but beware the current limitation.

I also didn't like how they show of their GraphQL compatibility and it's only when you look deep inside their doc that you figure out it's not really supported yet, felt like click bait.

**Comment by Vadorequest** [comment]
@jajourda I've opened https://github.com/UnlyEd/next-right-now/issues/18 and I could use your feedback!

**Comment by joshualinog** [comment]
@Vadorequest , i just dumped a response on #18  😄

**Comment by Vadorequest** [comment]
@jajourda Just discovered https://docs.directus.io/, looks very very good.

**Comment by joshualinog** [comment]
checking out directus.
We have repeatedly discussed GraphQL.

https://docs.directus.io/api/graphql.html

I posted screenshots below. How does this factor into your plans for the project and incorporating a GraphQL approach? Is it workable?

![image](https://user-images.githubusercontent.com/1839603/77922689-5a4aa880-726f-11ea-85a3-afa415d5533f.png)
![image](https://user-images.githubusercontent.com/1839603/77922703-5fa7f300-726f-11ea-9c85-26863b267c09.png)

**Comment by joshualinog** [comment]
Looking at docs right now on direcuts. It looks promising. I'm curious to see what plans you have for implementation in next-right-now. maybe:  **nrn_free_variant**? 😄

**Comment by Vadorequest** [comment]
I've gone a bit deeper regarding the Directus GraphQL capabilities and limitations.

Basically, they don't handle permissions (yet) when using GraphQL, so you have the ability to fetch "any" data. 

- I don't know yet what that means for sensitive data, I haven't looked into it and don't know how the GraphQL schema is generated. Does that mean we can fetch user passwords? I can't answer that yet.
- You can fetch all the data, so it's not a limitation if working with public data, only when using sensitive data it becomes a concerns
- There is an alternative RESTful API available anyway, with proper permissions support
- The GraphQL implementation is fairly new and being improved
- Overall, I'm not concerned by the GraphQL current limitations, even if GraphQL were not suitable there is still the REST API alternative

![image](https://user-images.githubusercontent.com/3807458/77939926-29429600-72b8-11ea-8d36-83e1f6b0e3e1.png)

And the name would rather be something like `nrn-v1-ssr-directus-no-graphcms`, but that doesn't really matter ATM ;)

**Comment by joshualinog** [comment]
I look forward to your implementation!

**Comment by joshualinog** [comment]
#### i have been working through my walkthrough on the `jajourda-docs` branch in my forked repo.

What would it be like to convert an existing _v1-ssr_ branch to the _free-preset_ that employs directus? I'm excited to try to employ the dircectus solution, but I want to wait for **best practices** if they are currently in the works.

**Comment by Vadorequest** [comment]
The v1-ssr and the current master branch are identical, I haven't started any preset for real at this time.

Best practices would be to remove unused features, by following "How to remove X" guides, I'm currently working on those best practices/guides.

But, to build a preset, you'd need strong and deep knowledge about how the boilerplate currently works of the preset you're starting from. My plan is to build a few presets that will remove one feature at a time, so that they may be used again to build other presets. Eventually, we'll end-up with some "simplest" preset with almost no 3rd party vendor built-in. And I'll use this "simplest" preset as quick start.

Hope that somehow answers your question.

**Comment by samuelcastro** [comment]
For an open source GraphQL server I'd use Hasura (https://hasura.io/) and just host it on Heroku for free while developing and then move it somewhere else afterwards.

**Comment by Vadorequest** [comment]
Interesting @samuelcastro, I've never used it myself but heard of it quite a few times. It seems it connects to an existing DB. What would be your recommandation to host such a DB for free as well? 

The goal being that the demo requires no upfront cost, whether related to GraphQL, servers or DB, etc.

**Comment by samuelcastro** [comment]
So you can host DB for free on Heroku as well for development, for production I'd use something like Amazon RDS which also offers some free piers when starting.

**Comment by samuelcastro** [comment]
Just to add more color, Hasura implements a patter called 3Factor (https://3factor.app/)

**Comment by Vadorequest** [comment]
Amazon RDS brings a ton of complexity, I wouldn't recommend it for getting started with a POC. It's very complex.

It's definitely doable, but I'm concerned it would bring much more complexity than what it'd solve.

On the other hand, hosting a DB for free on Heroku is very interesting!

**Comment by Vadorequest** [comment]
@jajourda I've tried Directus but I'm not sure if I'll make an example with it. It was complicated (IMO) to begin with.

The goal isn't necessarily to use GraphQL, and I believe my next attempt will be based on Airtable (or similar).

I'm currently heavily working on #42 but once it's done I believe I'll focus on a airtable preset.

**Comment by Vadorequest** [comment]
@jajourda FYI https://github.com/UnlyEd/next-right-now/issues/71

**Comment by Vadorequest** [comment]
FYI GraphCMS has increased the capacity of its free plan today. (x5 records)
https://graphcms.com/pricing

![image](https://user-images.githubusercontent.com/3807458/86388488-2b403a80-bc95-11ea-8cec-680dc234b9d0.png)

**Comment by smblee** [comment]
I second looking into hasura.
